syntax = "proto3";
package kv;

service KeyValue {
  rpc Put(PutRequest) returns (PutResponse);
  rpc Get(GetRequest) returns (GetResponse);
  rpc Replicate(PutRequest) returns (PutResponse);
  rpc GetChunkHash(ChunkRequest) returns (ChunkHashResponse);
  rpc FetchRange(RangeRequest) returns (stream KeyValuePair);
}

message PutRequest {
  string key = 1;
  string value = 2;
  int64 modified_at = 3;  // timestamp for conflict resolution
}
message PutResponse { bool ok = 1; string message = 2; }
message GetRequest { string key = 1; }
message GetResponse { string value = 1; bool found = 2; }
message KeyValuePair { string key = 1; string value = 2; int64 modified_at = 3; }
message ChunkRequest { int32 chunk_id = 1; }
message ChunkHashResponse { bytes hash = 1; }
message RangeRequest { int32 chunk_id = 1; }